File: applications/external/vgm_library/engine/Arduino/vgm_game_engine/src/internal/image.cpp
#include "image.h"
#include <stdint.h>
#include <string.h>

namespace VGMGameEngine
{

    void setup_fs()
    {
        if (!LittleFS.begin())
        {
            if (LittleFS.format())
            {
                LittleFS.begin();
            }
            else
            {
                rp2040.reboot();
            }
        }
    }

    bool EasyFile::open(const char *path, const char *mode)
    {
        this->file = LittleFS.open(path, mode);
        return this->file;
    }

    String EasyFile::read(const char *path)
    {
        String data = "";
        if (this->open(path, "r"))
        {
            data = this->file.readString();
            this->file.close();
        }
        return data;
    }

    uint8_t *EasyFile::read_bytes(const char *path, size_t &size)
    {
        uint8_t *data = nullptr;
        if (this->open(path, "r"))
        {
            size = this->file.size();
            data = new uint8_t[size];
            this->file.read(data, size);
            this->file.close();
        }
        return data;
    }

    bool EasyFile::write(const char *path, String data)
    {
        if (this->open(path, "w"))
        {
            this->file.print(data);
            this->file.close();
            return true;
        }
        return false;
    }

    // Destructor: free any allocated memory.
    Image::~Image()
    {
        if (buffer)
        {
            delete[] buffer;
            buffer = nullptr;
        }
        if (data)
        {
            delete[] data;
            data = nullptr;
        }
    }

    // from_path: load the BMP file and create the image buffer.
    bool Image::from_path(const char *path)
    {
        if (!open_image(path))
        {
            return false;
        }
        if (!create_image_buffer())
        {
            return false;
        }
        return true;
    }

    // from_byte_array: create image buffer from an external byte array.
    bool Image::from_byte_array(uint8_t *srcData, Vector newSize)
    {
        if (srcData == nullptr)
            return false;
        this->size = newSize;

        if (!this->is_8bit)
        {
            uint32_t numPixels = size.x * size.y;
            buffer = new uint16_t[numPixels];
            if (!buffer)
                return false;

            // Convert uint8_t array to uint16_t array (little endian)
            for (uint32_t i = 0; i < numPixels; i++)
            {
                buffer[i] = ((uint16_t)srcData[2 * i + 1] << 8) | srcData[2 * i];
            }

            if (board.libraryType == LIBRARY_TYPE_PICO_DVI)
            {
                // write in big-endian
                for (uint32_t i = 0; i < numPixels; i++)
                {
                    uint16_t pixel = buffer[i];
                    buffer[i] = (pixel >> 8) | (pixel << 8);
                }
            }
        }
        else
        {
            // 8-bit image: just copy the data directly into the data buffer.
            // then we'll use the data to draw bitmaps in level.cpp
            uint32_t numPixels = size.x * size.y;
            data = new uint8_t[numPixels];
            if (!data)
                return false;
            // Copy the data from the source to the new buffer.
            memcpy(data, srcData, numPixels);
        }

        return true;
    }

    // from_byte_array: create image buffer from an external byte array.
    bool Image::from_byte_array(const uint8_t *srcData, Vector newSize)
    {
        if (srcData == nullptr)
            return false;
        this->size = newSize;

        if (!this->is_8bit)
        {
            uint32_t numPixels = size.x * size.y;
            buffer = new uint16_t[numPixels];
            if (!buffer)
                return false;

            // Convert uint8_t array to uint16_t array (little endian)
            for (uint32_t i = 0; i < numPixels; i++)
            {
                buffer[i] = ((uint16_t)srcData[2 * i + 1] << 8) | srcData[2 * i];
            }

            if (board.libraryType == LIBRARY_TYPE_PICO_DVI)
            {
                // write in big-endian
                for (uint32_t i = 0; i < numPixels; i++)
                {
                    uint16_t pixel = buffer[i];
                    buffer[i] = (pixel >> 8) | (pixel << 8);
                }
            }
        }
        else
        {
            // 8-bit image: just copy the data directly into the data buffer.
            // then we'll use the data to draw bitmaps in level.cpp
            uint32_t numPixels = size.x * size.y;
            data = new uint8_t[numPixels];
            if (!data)
                return false;
            // Copy the data from the source to the new buffer.
            memcpy(data, srcData, numPixels);
        }

        return true;
    }

    // from_string: parse a string where each non-space character maps to a color.
    bool Image::from_string(String strData)
    {
        // Temporary storage for the pixel values (as 16-bit values)
        const int maxPixels = 1024 * 1024; //  maximum image size
        uint16_t *pixels = new uint16_t[maxPixels];
        if (!pixels)
            return false;

        int width = 0;
        int height = 0;
        int lineWidth = 0;

        // Process the string character by character.
        for (int i = 0; i < strData.length(); i++)
        {
            char c = strData.charAt(i);
            if (c == '\n')
            {
                // End-of-line: update width (only from the first row) and count the line.
                if (lineWidth > 0)
                {
                    if (height == 0)
                    {
                        width = lineWidth;
                    }
                    height++;
                    lineWidth = 0;
                }
            }
            else if (c == ' ')
            {
                // Skip spaces.
            }
            else
            {
                uint16_t pixel = 0;
                switch (c)
                {
                case '.':
                case 'f':
                    pixel = 0;
                    break;
                case '1':
                    pixel = 0xFFFF;
                    break;
                case '2':
                    pixel = 0xF904;
                    break;
                case '3':
                    pixel = 0xFC98;
                    break;
                case '4':
                    pixel = 0xFC06;
                    break;
                case '5':
                    pixel = 0xFFA1;
                    break;
                case '6':
                    pixel = 0x24F4;
                    break;
                case '7':
                    pixel = 0x7ECA;
                    break;
                case '8':
                    pixel = 0x0215;
                    break;
                case '9':
                    pixel = 0x879F;
                    break;
                case 'a':
                    pixel = 0xC05E;
                    break;
                case 'b':
                    pixel = 0xFC9F;
                    break;
                case 'c':
                    pixel = 0x50CA;
                    break;
                case 'd':
                    pixel = 0xACF0;
                    break;
                case 'e':
                    pixel = 0x7B07;
                    break;
                default:
                    pixel = 0x0020;
                    break;
                }
                pixels[height * width + lineWidth] = pixel;
                lineWidth++;
            }
        }
        // If the last line does not end with a newline, count it.
        if (lineWidth > 0)
        {
            if (height == 0)
            {
                width = lineWidth;
            }
            height++;
        }

        // Verify that we read some pixels.
        if (width <= 0 || height <= 0)
        {
            delete[] pixels;
            return false;
        }

        // Store size and allocate the buffer.
        this->size = Vector(width, height);
        uint32_t bufSize = width * height * 2;
        buffer = new uint16_t[width * height];
        if (!buffer)
        {
            delete[] pixels;
            return false;
        }

        // Copy the pixel values into the buffer.
        // (No vertical flip is done here because the string data is assumed to be top-down.)
        for (int i = 0; i < width * height; i++)
        {
            buffer[i] = pixels[i];
        }

        delete[] pixels;
        return true;
    }

    // open_image: opens a 16-bit BMP file (via EasyFile) and reads its header and pixel data.
    // This implementation reads the entire file into memory and then parses the BMP header.
    // Only uncompressed 16-bit BMP files are supported.
    bool Image::open_image(const char *path)
    {
        EasyFile file;
        size_t fileSize = 0;
        uint8_t *fileData = file.read_bytes(path, fileSize);
        if (!fileData)
        {
            return false;
        }

        // Basic check for BMP header size.
        if (fileSize < 54)
        {
            delete[] fileData;
            return false;
        }

        // Check BMP signature ("BM")
        if (!(fileData[0] == 'B' && fileData[1] == 'M'))
        {
            delete[] fileData;
            return false;
        }

        // File size (bytes 2-5) (unused here)
        uint32_t bmpFileSize = fileData[2] | (fileData[3] << 8) | (fileData[4] << 16) | (fileData[5] << 24);

        // Data offset (bytes 10-13): where pixel data starts.
        uint32_t dataOffset = fileData[10] | (fileData[11] << 8) | (fileData[12] << 16) | (fileData[13] << 24);

        // Info header size (bytes 14-17) (unused here)
        uint32_t infoHeaderSize = fileData[14] | (fileData[15] << 8) | (fileData[16] << 16) | (fileData[17] << 24);

        // Image width and height (bytes 18-25)
        int32_t width = fileData[18] | (fileData[19] << 8) | (fileData[20] << 16) | (fileData[21] << 24);
        int32_t height = fileData[22] | (fileData[23] << 8) | (fileData[24] << 16) | (fileData[25] << 24);
        this->size = Vector(width, height);

        // Skip planes (bytes 26-27); read bit depth at bytes 28-29.
        uint16_t depth = fileData[28] | (fileData[29] << 8);
        if (depth != 16)
        {
            delete[] fileData;
            return false;
        }

        // Compression (bytes 30-33) – for now we assume uncompressed or use compression value as needed.
        uint32_t compression = fileData[30] | (fileData[31] << 8) | (fileData[32] << 16) | (fileData[33] << 24);
        if (compression != 0)
        {
            delete[] fileData;
            return false;
        }

        // Calculate the row size (each row is padded to a multiple of 4 bytes).
        uint32_t rowSize = ((width * 2 + 3) / 4) * 4; // each pixel is 2 bytes

        // Allocate our data buffer for pixel values (width * height * 2 bytes).
        data = new uint8_t[width * height * 2];
        if (!data)
        {
            delete[] fileData;
            return false;
        }

        // BMP files store rows from bottom to top.
        // Copy each row (ignoring padding) into our data buffer.
        for (int row = 0; row < height; row++)
        {
            // In the file, row0 is the bottom row.
            uint8_t *srcRow = fileData + dataOffset + row * rowSize;
            // We store rows in the same order as read (i.e., bottom-up).
            memcpy(data + row * width * 2, srcRow, width * 2);
        }

        delete[] fileData;
        return true;
    }

    bool Image::create_image_buffer()
    {
        if (!data)
            return false;

        // Get image dimensions as integers.
        int width = (int)size.x;
        int height = (int)size.y;

        uint32_t numPixels = width * height;
        buffer = new uint16_t[numPixels];
        if (!buffer)
            return false;

        // Reinterpret pointers as 16-bit arrays.
        uint16_t *buf16 = buffer;
        uint16_t *data16 = (uint16_t *)data;

        // Flip rows: BMP data is stored bottom-up and we want top-down.
        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                buf16[y * width + x] = data16[(height - 1 - y) * width + x];
            }
        }

        return true;
    }

}

File: src/MicroPython/src/pico_game_engine/image.py
# from https://github.com/harlepengren/PicoGame/blob/main/picogame/image.py
# edited by JBlanked on 2025-01-19
import framebuf
import array
import sys
from .vector import Vector

SEEK_CUR = 1


class Image:
    """
    Represents an image.

    The image is stored in a buffer that can be drawn to the screen.
    You can create an image from a file path, a byte array, or a string.
    If you create an image from a file, the image must be in 16-bit BMP format.
    """

    def __init__(self):
        self.size = Vector(0, 0)
        self.buffer = None
        self._image_data = []

    def from_path(self, path: str):
        """Create an image from a file path"""
        self.open_image(path)
        self.create_image_buffer()

    def from_byte_array(self, data, width, height):
        """
        Create an image from a byte array, matching the C++ version.

        Args:
            data (bytes or bytearray): The byte array containing pixel data.
            width (int): The width of the image.
            height (int): The height of the image.

        Raises:
            ValueError: If the length of data does not match width * height * 2.
        """
        # Validate data length
        expected_length = width * height * 2
        if len(data) != expected_length:
            raise ValueError(
                f"Data length {len(data)} does not match expected size {expected_length} (width={width}, height={height})!"
            )

        self.size = Vector(width, height)

        # Convert byte array to array of unsigned shorts ('H') in little endian
        pixel_array = array.array("H")
        pixel_array.frombytes(data)

        # Ensure the byte order is little endian
        if sys.byteorder != "little":
            pixel_array.byteswap()

        # Create FrameBuffer from the bytearray representation of pixel_array
        byte_buffer = pixel_array.tobytes()
        self.buffer = framebuf.FrameBuffer(byte_buffer, width, height, framebuf.RGB565)

        return True  # Indicate success

    def open_image(self, path):
        """Open a 16-bit BMP file and read the image data"""
        with open(path, mode="rb") as f:
            signature = f.read(2)
            fsize = int.from_bytes(f.read(4), "little")
            f.seek(4, SEEK_CUR)
            dataOffset = int.from_bytes(f.read(4), "little")

            infoHeaderSize = int.from_bytes(f.read(4), "little")
            self.size = Vector(
                int.from_bytes(f.read(4), "little"),
                int.from_bytes(f.read(4), "little"),
            )
            f.seek(2, SEEK_CUR)
            depth = int.from_bytes(f.read(2), "little")

            # If compression is 3 then we also need the mask fields to determine whether it is RGB555 or RGB565
            compression = int.from_bytes(f.read(4), "little")
            imageSize = int.from_bytes(f.read(4), "little")

            f.seek(4, SEEK_CUR)

            # colorUsed = struct.unpack('2I', f.read(8))
            f.seek(8, SEEK_CUR)
            masks = (
                int.from_bytes(f.read(4), "little"),
                int.from_bytes(f.read(4), "little"),
                int.from_bytes(f.read(4), "little"),
                int.from_bytes(f.read(4), "little"),
            )

            # Go to end of head
            f.seek(dataOffset)

            rowSize = int(self.size.x * depth / 32 * 4)

            self._image_data = []
            for currentRow in range(0, self.size.y):
                for currentColumn in range(0, self.size.x):
                    currentRead = f.read(2)
                    if len(currentRead) > 0:
                        self._image_data.append(currentRead)

    def create_image_buffer(self):
        """Create a frame buffer from the image data"""
        width = self.size.x
        height = self.size.y
        temp_buffer = bytearray(width * height * 2)
        self.buffer = framebuf.FrameBuffer(temp_buffer, width, height, framebuf.RGB565)

        # We need to flip rows because bmp starts on the bottom left.
        for y in range(0, height):
            for x in range(0, width):
                self.buffer.pixel(
                    x,
                    y,
                    int.from_bytes(
                        self._image_data[(height - 1 - y) * width + x], "little"
                    ),
                )

    def from_string(self, image_str: str):
        """Use a string to create an image. One byte per pixel. Height and width are inferred based on carriage returns.
        The input must have equal length rows."""

        width = 0
        height = 0

        self._image_data = []
        for _char in image_str:
            if _char == "\n":
                if height == 1:
                    # First line, let's get the width
                    width = len(self._image_data)
                height += 1
            elif _char == " ":
                pass
            elif (_char == ".") or (_char == "f"):
                self._image_data.append(0)
            elif _char == "1":
                self._image_data.append(int(0xFFFF))
            elif _char == "2":
                self._image_data.append(int(0xF904))
            elif _char == "3":
                self._image_data.append(int(0xFC98))
            elif _char == "4":
                self._image_data.append(int(0xFC06))
            elif _char == "5":
                self._image_data.append(int(0xFFA1))
            elif _char == "6":
                self._image_data.append(int(0x24F4))
            elif _char == "7":
                self._image_data.append(int(0x7ECA))
            elif _char == "8":
                self._image_data.append(int(0x0215))
            elif _char == "9":
                self._image_data.append(int(0x879F))
            elif _char == "a":
                self._image_data.append(int(0xC05E))
            elif _char == "b":
                self._image_data.append(int(0xFC9F))
            elif _char == "c":
                self._image_data.append(int(0x50CA))
            elif _char == "d":
                self._image_data.append(int(0xACF0))
            elif _char == "e":
                self._image_data.append(int(0x7B07))
            else:
                self._image_data.append(int(0x0020))

        height -= 1

        temp_buffer = bytearray(width * height * 2)
        self.buffer = framebuf.FrameBuffer(temp_buffer, width, height, framebuf.RGB565)

        # We need to flip rows because bmp starts on the bottom left.
        for y in range(0, height):
            for x in range(0, width):
                self.buffer.pixel(x, y, self._image_data[y * width + x])

        self.size = Vector(width, height)
